Give each status a copyright

mike-burns opened the pull request:
This PR is incomplete but what do we think of the idea?

- Works of art have a copyright in most countries; this documents the currently-implied copyright.
- This could set the way for changing the copyright on a per-status basis, if we were to expose that in the UI.

Also, I hate the string duplication but wanted to get a discussion going before I figured out how to best fix that within this codebase. Same with the lack of tests.

---

Every work created by a person has a copyright of some type, in most
countries. Document the previously-implied copyright using the Creative
Commons license Attribution 4.0 International (CC BY 4.0).

https://creativecommons.org/licenses/by/4.0/

For ActivityPub/ActivityStreams, bring in the ccREL ontology, connecting
the work with a `license` property.

For OStatus, augment the Atom entry with a link to the license, as
documented by CC: https://wiki.creativecommons.org/wiki/Atom

Add a `license` value to the REST API for anything involving a status.

This does not expose anything user-facing; it simply seeks to document
the current state of things in a machine-readable way.

mike-burns committed:
Give each status a copyright

Every work created by a person has a copyright of some type, in most
countries. Document the previously-implied copyright using the Creative
Commons license Attribution 4.0 International (CC BY 4.0).

https://creativecommons.org/licenses/by/4.0/

For ActivityPub/ActivityStreams, bring in the ccREL ontology, connecting
the work with a `license` property.

For OStatus, augment the Atom entry with a link to the license, as
documented by CC: https://wiki.creativecommons.org/wiki/Atom

Add a `license` value to the REST API for anything involving a status.

This does not expose anything user-facing; it simply seeks to document
the current state of things in a machine-readable way.

akihikodaki reviewed (changes requested):
The idea is great (one of my friends gives an explicit license for his posts and he would be grateful with the change), but I do not agree with the implementation.

First, your change will be applied to all the old posts when the server has the later versions. That means updating their instances will require to update the acknowledgement of _all_ users. That is nearly impossible for some instances.

Secondly, your change does not allow to change the license. Why doesn't it do so while it is allowed to change the instance policy?

At the end, I strongly recommend to have notices about the license somewhere on the UI and to provide an API to retrieve the license when posting. The lack of the feature gives doubts on the validity of the license claim, and even `cc:license` context may be considered as non-reliable information, which is my major concern.

akihikodaki commented:
For your information, the terms of Discourse, on which the terms of Mastodon is based, has section named _User Content License_, and says the contributions will be licensed under CC-BY-NC-SA 3.0. It is a nice way to tell users about their copyrights.

mike-burns commented:
Thank you for the feedback, @akihikodaki !

Here's how I'll move this forward:

- Add a `license` field to the `statuses` table.
- Provide a UI for selecting a license -- in the media/privacy/CW cluster seems right.
- Only send license information if one is explicitly selected for that status.
- Have a default license applied to the status, which can be saved for the user as a setting.
- Show license information on a status in the Web UI.

Seem right?

Some decisions that I don't feel qualified to make:

- This is my first time working with OStatus, ActivityPub, etc. Did I get all the parts of the codebase where we send the status around?
- What is the default license that new instances should ship with?  Is CC attribution-only correct? Public domain? None? Again this gets into the weird territory where all works do have a copyright so "none" (in most countries) is not actually a legally-correct option.
- Is it OK to have a drop-down pre-populated with all the CC licenses? Should it be free-form text?
- People may want to retroactively apply licenses to prior statuses -- should we have a UI for that?
- I am mixing the words "license" and "copyright" in here -- which is this, and is it OK to conflate them a little bit?

akihikodaki commented:
@mike-burns Thank you for your quick reply.

> Here's how I'll move this forward:

They look nice, but also tough to implement. I would highly appreciate if you made those changes. Here is some feedback for them.

> * Provide a UI for selecting a license -- in the media/privacy/CW cluster seems right.

I don't prefer to add a more control to the UI. The column is messy even now. It is still fine to expose an API for third-party applications to provide the same feature. Please note that this is just my opinion and there is no clear convention about new change for the column, so others may have different opinions.

> * Only send license information if one is explicitly selected for that status.

I want to make sure what a license "explicitly selected" would be. In my opinion:
* The statuses (i) made by users who have registered after their instances applied this change should have an explicitly selected license because such instances will have a default license and have told them about that.

* The statuses (i) made by users who have registered before their instances applied this change and (ii) made before they set their default license (iii) whose license have not selected via the compose UI will _not_ have an explicitly selected license.

Technically, that can be implemented by:
* adding (possibly nullable) `default_license` field to the instance configuration,
* adding nullable `default_license` field to `users` table, and
* adding nullable `license` field to `statuses` table.

The `license` field for statuses made before this change should be `NULL`.
Otherwise, the `license` field should be:
* the value chosen in the compose form if one is chosen
* or `default_license` field of the user if it is not `NULL`
* or `default_license` field of the instance

If the `license` field of a status is `NULL`, the status does not have its explicitly selected license. Otherwise, the value indicates its explicitly selected license.

> This is my first time working with OStatus, ActivityPub, etc. Did I get all the parts of the codebase where we send the status around?

It looks good to me. However, please note that you may add a change to interpret the license field of remote statuses and fill the corresponding column of the database if you are adding a change for the database.

> What is the default license that new instances should ship with? Is CC attribution-only correct? Public domain? None? Again this gets into the weird territory where all works do have a copyright so "none" (in most countries) is not actually a legally-correct option.

The following is just my preference and other reviewers may have different opinion, but:
* I prefer the latest CC-BY-SA.
* Public domain is not good. It may introduce a kind of stereotype that everybody can do anything for statuses on Mastodon. At least we need to show that we are taking account of copyrights.
* "None", is actually safe because it is the current "default" license.

Again, thank you for your considerations.

akihikodaki commented:
I have not answered all questions...

> Is it OK to have a drop-down pre-populated with all the CC licenses? Should it be free-form text?

Free-form text does not seem possible because of the specification of CC. I suggest to implement with pre-populated CC licenses as the initial version.

> People may want to retroactively apply licenses to prior statuses -- should we have a UI for that?

I do not think so because:
* updating federated statuses is tough and
* usually you cannot withdraw the license

* I am mixing the words "license" and "copyright" in here -- which is this, and is it OK to conflate them a little bit?

I understand those words are different so you do not have to. You may choose appropriate one. Here is my long description: (disclaimer: I am not a lawyer and I am Japanese.)
In my understanding, copyright is the right of the owner and license is statement describing the condition the owner does _not_ begin disputes. So you may say you give a license for each status and you will not sue others using your copyright while they are conforming to the license. Note that copyright is not necessarily "given"; but license is.

mike-burns committed:
Add default_license setting

Note that `Setting::LICENSES` does not belong there but that's where it
will have to live for now.

mike-burns committed:
Store the license on the Status, send it in JSON

This moves `Setting::LICENSES` to `Status::LICENSE_URLS`.

Only present the license URL if the status has a license. (On the
OStatus; unsure about ActivityPub and REST.)

Store the license on the Status so that it is persisted even if the user
changes their default. This also paves the way for one-off licenses.

mike-burns committed:
Backfill tests

While here: only set the license if it is unset. This makes tests easier
and also makes future expansion easier.

mike-burns committed:
Expose the license in the Web UI

Send both the `license` and `license_url` so we can link nicely.

mike-burns committed:
Code style consistency

mike-burns commented:
I found a few hours so here's an update:

- Adds a `default_license` setting on the user, with a dropdown in the settings presenting all the CC licenses.
- The `default_license`is applied to new statuses, persisted on the `statuses` table.
- The status' license URL is sent via OStatus, ActivityPub, and the REST API.
- The REST API also sends the license name.
- Show the license in the Web UI (JS and HTML, with some CSS).
- If the default license is unset, ignore it.
- The the status' license is unset, ignore it.
- Some light testing.

Still on my to-do list:
- Instance defaults. Or can this be merged without that feature?

I need help:

- How should it look in the Web and JS UI? If it should be any fancier than this, I'd love assistance there.
- Are there tests for the serializers that I had missed?
- Are there JS tests? Integration/system/feature tests?
- Translations.

Is this a fine start?

mike-burns commented:
@trwnh I see you gave this a thumbs-down. Anything you want to discuss?

mike-burns committed:
Remove unnecessary license_url delegation

mike-burns commented on app/views/settings/preferences/show.html.haml:
```diff
@@ -18,6 +18,8 @@
 
     = f.input :setting_default_sensitive, as: :boolean, wrapper: :with_label
 
+    = f.input :setting_default_license, collection: Status::LICENSE_URLS.keys, wrapper: :with_label
```
Perhaps `Status::LICENSE_URLS.keys` is not user-friendly. Maybe this whole input is confusing. Very open to opinions here.

akihikodaki replied:
It also prevents from localization. Creative Commons localizes license names on their Web sites. It is better to conform the prior instances and localize license names as they do.

trwnh commented:
@mike-burns Not particularly. I just disagree with implementing anything that reifies copyrights in general, especially in open software.

akihikodaki commented:
> Not particularly. I just disagree with implementing anything that reifies copyrights in general, especially in open software.

The copyright is based on laws and cannot be ignored by any means. If you disagree the idea of copyrights, you need to clarify the stance (that is why FSF made GPL.)

akihikodaki commented on app/models/status.rb:
```diff
@@ -31,6 +31,15 @@ class Status < ApplicationRecord
+  LICENSE_URLS = {
+    'Attribution (CC BY)' => 'https://creativecommons.org/licenses/by/4.0/',
+    'Attribution-ShareAlike (CC BY-SA)' => 'https://creativecommons.org/licenses/by-sa/4.0/',
+    'Attribution-NoDerivs (CC BY-ND)' => 'https://creativecommons.org/licenses/by-nd/4.0/',
```
The web page says "Attribution-NoDerivatives" (while it says "Attribution-NonCommercial-NoDerivs" for BY-NC-ND. How confusing!)

akihikodaki commented on app/serializers/activitypub/note_serializer.rb:
```diff
@@ -86,6 +86,10 @@ def local?
     object.account.local?
   end
 
+  def license
```
Use `key` option of `attribute` method. See http://www.rubydoc.info/gems/active_model_serializers/0.10.6/ActiveModel/Serializer#attribute-class_method for details.

akihikodaki reviewed (changes requested):
Because of localization of license names, I'd like to have changes to use identifiers which are different from the names for the database. URIs are totally fine in my opinion.

trwnh commented:
@akihikodaki:
> The copyright is based on laws and cannot be ignored by any means. If you disagree the idea of copyrights, you need to clarify the stance (that is why FSF made GPL.)

To clarify, I do not think that copyright status should be supported in general. It may be the law in several countries, but the existence of laws does not require accommodating them. I would oppose anything that made it easier to enforce that law. In this case, that would mean *not* providing an exposed copyright status in the UI. This is my belief, and I am expressing it here to provide some dissent to the idea, though this pull request may ultimately end up being accepted.

mike-burns commented:
Ah great points re localization. I'll change what we store in the DB and how that makes it to the user. I'll also make the other improvements you suggest.

Is this idea OK to merge without instance defaults? As-is it will preserve existing functionality. We can add instance defaults in a separate diff.

akihikodaki commented:
@trwnh
> I would oppose anything that made it easier to enforce that law.

That makes sense.

@mike-burns

> Ah great points re localization. I'll change what we store in the DB and how that makes it to the user. I'll also make the other improvements you suggest.

Nice!

> Is this idea OK to merge without instance defaults? As-is it will preserve existing functionality.

Yes.

> We can add instance defaults in a separate diff.

I agree.

mike-burns committed:
Small improvements

mike-burns committed:
Store the license_url, not the license name

Instead of storing an English string describing the license, we can
store the URL to the license. Provide the English string via i18n.

We cannot use the URL as the index into i18n because it contains a
period, so provide a mapping back to a string that we can then use to
index into the i18n backend.

mike-burns committed:
Syntax

Those values no longer need to be symbols because simple_form doesn't
respect them anyway.

mike-burns committed:
Populate translations

This uses the text from creativecommons.org. When that did not provide a
translation, use English.

mike-burns commented:
@akihikodaki updated:

- Store the URL in the DB.
- Use i18n to show the label.
- Populate the translations from creativecommons.org.

[I typically work on this branch in 1 hour bursts but today I had the time to spend the requisite number of hours remembering how simple_form, collections, and i18n interact.]

mike-burns committed:
Normalize and use all CC translations

akihikodaki reviewed (changes requested):
It is almost done! The only thing left is to implement receivers for federation. See `app/lib/activitypub/activity/create.rb` for ActivityPub, `app/lib/ostatus/activity/creation.rb` for OStatus.

akihikodaki commented:
@Gargron This pull request has grown much. I would like to know if there is a chance to get this merged.

mike-burns committed:
Consume status with licenses

When we receive, via OStatus or ActivityPub, a status with a license,
persist that in the `statuses` table.

This introduces a simple test for `OStatus::Activity::Post`. The fixture
for this path is tricky: the status must have an id, but it cannot be
persisted in our database. So create it, generate the XML, then destroy
it.

mike-burns commented:
Ah right good catch; added to the receiver side of things.

Much of the diff growth is in i18n.

mike-burns committed:
Style

mike-burns commented:
If you'd like me to re-write some git history just say the word -- though let's wait until we even know that we got it right and that we want it merged.

akihikodaki commented on app/lib/ostatus/atom_serializer.rb:
```diff
@@ -348,6 +348,10 @@ def add_namespaces(parent)
     append_element(entry, 'link', nil, rel: :alternate, type: 'application/activity+json', href: ActivityPub::TagManager.instance.uri_for(status)) if status.account.local?
 
+    if status.license_url
+      append_element(entry, 'link', rel: :license, type: 'application/rdf+xml', href: status.license_url)
```
It should not say `type: 'application/rdf+xml'`.

I noticed that `LICENSE_URLS` are not for `application/rdf+xml`. For example, the RDF for CC BY is not `https://creativecommons.org/licenses/by/4.0/`, but `https://creativecommons.org/licenses/by/4.0/rdf`.
So either `LICENSE_URLS` or the `type` attribute must be changed. According to CC REL specification, removing the `type` attribute is better.

akihikodaki reviewed (changes requested):
The OStatus interpreter tests if the `type` attributes of links are `application/rdf+xml`, but it seems incorrect.

akihikodaki commented:
The Git history is fine. We do not have any rule to define how to organize commits as far as I know.

mike-burns committed:
Relax the license link constraint

All we care is that it links to a license, not the media type of the
license URL.

cwebber commented on app/lib/activitypub/adapter.rb:
```diff
@@ -17,6 +17,8 @@ class ActivityPub::Adapter < ActiveModelSerializers::Adapter::Base
         'toot'                      => 'http://joinmastodon.org/ns#',
         'Emoji'                     => 'toot:Emoji',
+        'cc'                        => 'https://creativecommons.org/ns#',
+        'license'                   => 'cc:license',
```
The license term doesn't exist in cc: (it's in fact the one thing not in cc: in ccRel), it's in the xhtml namespace.  Since you probably only need one term from xhtml I'd simply do:

```
 'license'       => 'http://www.w3.org/1999/xhtml/vocab#license'
```

Gargron commented on app/serializers/rest/status_serializer.rb:
```diff
@@ -3,7 +3,8 @@
              :sensitive, :spoiler_text, :visibility, :language,
-             :uri, :content, :url, :reblogs_count, :favourites_count
+             :uri, :content, :url, :reblogs_count, :favourites_count,
+             :license, :license_url
```
It feels like it would be a bit cleaner, API wise, to have a `license: { name: ..., url: ... }` object

Gargron reviewed (changes requested):
With a couple changes as requested by me, aki and cwebber I think this is a viable addition. Sorry it took like, 8 months to get to this, time flies and it feels like new year's was just yesterday.

nightpool commented:
Two questions on UX:

 - After a user selects a default license, how do they unset it?
 - Can the user put in a custom default license value? Do we think they should be able to?

mike-burns commented on db/migrate/20171221044254_add_license_to_statuses.rb:
```diff
@@ -0,0 +1,5 @@
+class AddLicenseToStatuses < ActiveRecord::Migration[5.1]
+  def change
+    add_column :statuses, :license_url, :text, null: true
```
For Postgres, this is the canonical article: https://www.depesz.com/2010/03/02/charx-vs-varcharx-vs-varchar-vs-text/

I'm willing to do something else if we want to support other DBMSes more natively?

Gargron replied:
We don't plan on supporting any other DB type since we already use some postgres-specific syntax and column types. However, no other URL has a "text" type so far and I don't see a reason to do this one differently.

mike-burns replied:
Cool -- I value consistency. I'll make the change.

mike-burns commented:
Thanks for the great feedback everyone. I'll rebase and address your comments in a few days.

Gargron commented on app/serializers/rest/status_serializer.rb:
```diff
@@ -86,6 +87,13 @@ def pinnable?
+    I18n.t(
+      Status::LICENSE_URLS[object.license_url],
+      scope: 'simple_form.options.user.setting_default_license'
+    )
```
Hmm. The only concern is, how do you deal with unrecognized licenses. As far as *recognized* licenses go, this seems fine. (Unless you want to push this concern to the app's i18n...)

nightpool replied:
i would just imagine that we store/federate license name along with license url

nightpool replied:
i guess that's more difficult though, and fucks up i18n

mike-burns commented:
@nightpool 

> After a user selects a default license, how do they unset it?

Currently not possible. Is that a feature we want?

One thing I keep coming back to in my mind is: every status has a license (in Europe and NA, at least); this is about exposing it and giving the user the option to change it.

But if you think it makes sense to, I can add a null value to the dropdown in `preferences/show.html.haml`.

> Can the user put in a custom default license value? Do we think they should be able to? does the additional nuance justify the additional UX complexity?

You raise some good points, especially technical ones inline. Regardless of how we limit the user on Mastodon there's also the rest of the fediverse to consider.

How should we federate the human-visible value of the license? Should we additionally send a default name (e.g. "BY-NC-SA") and leave it up to the client to render that as they please (we would continue to run it through i18n (e.g. "Attribution-NonCommercial-ShareAlike (CC BY-NC-SA)"))? Should we continue to send only the URL and show something like "©" with a link to their URL if we don't recognize it? And what of licenses that don't have a URL?

Simplest option: we only support CC licenses for now. Only send URL, have a fallback for URLs we don't recognize.

Happy to change this as you see fit. Let me know how you think we can make this work.

mike-burns committed:
Changes from review

- Send a license object in the REST API.
- Drop the `type` from the `link` element for the license in OStatus,
since we do not necessarily know the type.
- Syntax consistency change: no comma at end of method arg list.
- DB consistency change: URLs are stored as `:string`, not `:text`.
